include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,301 @@
+		using state_type = CStateBase; //!< full combined state
+
+		static inline bool validate(state_type* state)
+		{
+			if (!CMatchedSizeInOutImageFilterCommon::validate(state))
+				return false;
+
+			if (!CSummedAreaTableImageFilterBase<ExclusiveMode>::validate(state))
+				return false;
+			
+			const ICPUImage::SCreationParams& inParams = state->inImage->getCreationParameters();
+			const ICPUImage::SCreationParams& outParams = state->outImage->getCreationParameters();
+			const auto inFormat = inParams.format;
+			const auto outFormat = outParams.format;
+
+			if (state->scratchMemoryByteSize < state_type::getRequiredScratchByteSize(state->inImage, state->outImage))
+				return false;
+
+			if (asset::getFormatChannelCount(outFormat) != asset::getFormatChannelCount(inFormat))
+				return false;
+
+			if (asset::getFormatClass(inFormat) >= asset::getFormatClass(outFormat)) // TODO in future! Change to a function checking a bit-depth for an each single channel
+				return false;
+
+			return true;
+		}
+
+		static inline bool execute(state_type* state)
+		{
+			if (!validate(state))
+			{
+				state->imageTotalCacheScratch = nullptr;
+				state->imageTotalCacheOutput = nullptr;
+				return false;
+			}
+
+			auto checkFormat = state->inImage->getCreationParameters().format;
+			if (isIntegerFormat(checkFormat))
+				return executeInterprated(state, reinterpret_cast<uint64_t*>(state->scratchMemory));
+			else
+				return executeInterprated(state, reinterpret_cast<double*>(state->scratchMemory));
+		}	
+
+	private:
+
+		template<typename decodeType> //!< double or uint64_t
+		static inline bool executeInterprated(state_type* state, decodeType* scratchMemory)
+		{
+			const asset::E_FORMAT inFormat = state->inImage->getCreationParameters().format;
+			const asset::E_FORMAT outFormat = state->outImage->getCreationParameters().format;
+			const auto inTexelByteSize = asset::getTexelOrBlockBytesize(inFormat);
+			const auto outTexelByteSize = asset::getTexelOrBlockBytesize(outFormat);
+			const auto currentChannelCount = asset::getFormatChannelCount(inFormat);
+			static constexpr auto maxChannels = 4u;
+			const auto outScratchLayerOffset = state->extent.width * state->extent.height * state->extent.depth * currentChannelCount;
+
+			uint8_t* const totalImageCacheScratch = state->imageTotalCacheScratch = state->scratchMemory + state->cachesOffset;
+			uint8_t* const totalImageCacheOutput = state->imageTotalCacheOutput = state->imageTotalCacheScratch + state->imageTotalScratchCacheByteSize;
+			decodeType* const mainRowCacheScratch = reinterpret_cast<decodeType*>(state->mainRowCacheScratch = state->imageTotalCacheOutput + state->imageTotalOutputCacheByteSize);
+			decodeType* const mainColumnCacheScratch = reinterpret_cast<decodeType*>(state->mainColumnCacheScratch = state->mainRowCacheScratch + state->mainRowScratchCacheByteSize);
+
+			memset(scratchMemory, 0, state->scratchMemoryByteSize);
```

Comment:
memset eats perf, allow undefined behaviour if images have undefined data in the extent you've asked for, this is what vulkan does anyway

Reply:
but leave it on for `IRR_DEBUG`

Reply:
corrected in following commit

Reply:
this debug thing should probably move to the per-layer loop